source/Halibut/Transport/ConnectionManagerAsync.cs

```diff
@@ -0,0 +1,300 @@
+                    }
+
+                    return existingConnectionFromPool;
+                }
+            }
+
+            var connection = await CreateNewConnectionWithIoAsync(exchangeProtocolBuilder, connectionFactory, serviceEndpoint, log, cancellationToken);
+
+            //Do as a separate lock so that we do not lock on IO from creating a new connection.
+            using (await connectionsLock.LockAsync(cancellationToken))
+            {
+                using (await activeConnectionsLock.LockAsync(cancellationToken))
+                {
+                    AddConnectionToActiveConnections(serviceEndpoint, connection);
+                }
+            }
+
+            return connection;
+        }
+        
+        [Obsolete]
+        Tuple<IConnection, Action> CreateNewConnection(ExchangeProtocolBuilder exchangeProtocolBuilder, IConnectionFactory connectionFactory, ServiceEndPoint serviceEndpoint, ILog log, CancellationToken cancellationToken)
+        {
+            var lazyConnection = new Lazy<IConnection>(() => connectionFactory.EstablishNewConnection(exchangeProtocolBuilder, serviceEndpoint, log, cancellationToken));
+            var connection = new DisposableNotifierConnection(lazyConnection, OnConnectionDisposed);
+            return Tuple.Create<IConnection, Action>(connection, () =>
+            {
+                // ReSharper disable once UnusedVariable
+                var c = lazyConnection.Value;
+            });
+        }
+        
+        async Task<IConnection> CreateNewConnectionWithIoAsync(ExchangeProtocolBuilder exchangeProtocolBuilder, IConnectionFactory connectionFactory, ServiceEndPoint serviceEndpoint, ILog log, CancellationToken cancellationToken)
+        {
+            var connection = await connectionFactory.EstablishNewConnectionAsync(exchangeProtocolBuilder, serviceEndpoint, log, cancellationToken);
+            return new DisposableNotifierConnection(new Lazy<IConnection>(() => connection), OnConnectionDisposed);
+        }
+
+        void AddConnectionToActiveConnections(ServiceEndPoint serviceEndpoint, IConnection connection)
+        {
+            if (activeConnections.TryGetValue(serviceEndpoint, out var connections))
+            {
+                connections.Add(connection);
+            }
+            else
+            {
+                connections = new HashSet<IConnection> {connection};
+                activeConnections.Add(serviceEndpoint, connections);
+            }
+        }
+
+        [Obsolete]
+        public void ReleaseConnection(ServiceEndPoint serviceEndpoint, IConnection connection)
+        {
+            using (connectionsLock.Lock())
+            {
+                pool.Return(serviceEndpoint, connection);
+                using (activeConnectionsLock.Lock())
+                {
+                    if (activeConnections.TryGetValue(serviceEndpoint, out var connections))
+                    {
+                        connections.Remove(connection);
```

Comment:
I wonder if a set of "activeConnection" methods could be made that take locks for the caller that way method only need to worry about taking connectionLocks.